### PR TITLE
refactor(opsem): make rawmemmgr available through mixin

### PR DESCRIPTION
**Pull request by priyasiddharth**
Moved MemRepr and Allocator interfaces to separate header files to make it easy to reason about cross dependencies since RawMemMgr, MemRepr and Allocator are strongly coupled together. 

TODO:
1. fold OpSemMemManagerBase into derived class since it is not used anywhere else
2. remove widememmanagermixin and use only one mixin
3. remove static eval_if decisions from extrawidememmanager such that these decisions are only taken in the mixin.

### CODE REVIEW

File: lib/seahorn/BvOpSem2ExtraWideMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2ExtraWideMemMgr.hh:44]: why these types degenerated to `Expr`?

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:26]: readability can be improved if throughout the code the `Mixin` are hidden under some defined type. So that lines like these are read as 'RawMemManager m_main` or similar.

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.hh:137]: `Expr` are smart pointers and are usually passed by value. This is an outlier.

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.hh:256]: `Expr` are passed by value

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:325]: `extract` and `concat` are now part of ALU. Need to update this code to use them.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:567]: different from previous behavior. Used to return `Expr(0)` when nothing was stored.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:697]: is this even type correct? Used to return null wrapped in an `Expr`, but now returns a potentially legal expression...

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:729]: very very different behavior. Previously used to return null that is interpreted as non-deterministic, when offset was not computed. Now returns a symbolic `nullptr`

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:782]: `RawMemManager::~RawMemManager() = default;`

**Review by agurfinkel — CHANGES_REQUESTED** [review]
looks good, but there are a couple of showstoppers that significantly change desired behavior. Once these are fixed, this can be merged

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.hh:256]: got it.Fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.hh:137]: got it.Fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc:782]: done

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc:729]: Good point. Fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc:325]: done

File: lib/seahorn/BvOpSem2TrackingRawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2TrackingRawMemMgr.hh:26]: done

File: lib/seahorn/BvOpSem2ExtraWideMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2ExtraWideMemMgr.hh:44]: Fixed... was trying something out during refactoring

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: Fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: Fixed.

**Comment by priyasiddharth** [comment]
Responded to review comments. PTAL.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:339]: for some reason, `zext`, `concat` and `extract` here are not used from `ALU`

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:439]: use `alu().Extract` instead of `bv::extract`

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:468]: `bv::extract` throughout this function

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:471]: `bv::concat`

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:496]: `zext` and `concat` and `bvnum`

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:731]: `zext`

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.hh:18]: is it `CoreMemManager` or `MemManagerCore` ? Would be nice to be consistent with `RawMemManagerCore` if the idea is that `Core` is the base of everything

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.hh:138]: no `const` if passed by value

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.hh:283]: note for future. `loadValueFromMem` and `storeValueToMem` can be implemented in the mixin because they are implemented completely in terms of other functions of the memory manager.

**Review by agurfinkel — COMMENTED** [review]
done another round. looks almost good. let me know when you are ready to merge

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.hh:18]: fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc:339]: fixed

**Comment by agurfinkel** [comment]
any update on this? I'm waiting to merge with my changes

**Comment by priyasiddharth** [comment]
Will send this in the next couple of hours

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.cc
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc]: fixed.

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.hh:283]: added note to mixin class

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.hh]: good point. fixed.

**Comment by priyasiddharth** [comment]
Ready for another round. RawMemManager should not have any leftover `alu` operations which bypass the `alu`. 

I have not squashed the commits into a single one. I can do that after getting an LGTM or you can rebase and merge and do it, whichever you prefer.

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.hh:118]: still an `Expr`. Is that an oversight?

**Review by agurfinkel — APPROVED** [review]
Ready to merge, squash if you want, and please merge

File: lib/seahorn/BvOpSem2RawMemMgr.hh
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.hh:118]: This doesn't use `OpSemMemManager` anymore so I didn't want to depend on types exported by that class. Since it was just redefining type, I took the actual definition directly. 

In the future, maybe we can have a class which encapsulates these types.

**Comment by priyasiddharth** [comment]
Pushing inspite of azure failure since it is a timeout which appears unrelated to these changes.

**Pull request merged**
